### PR TITLE
pd-client: implement GetTSAsync use streaming

### PR DESCRIPTION
**Pull request by tiancaiamao**
Re-implement of https://github.com/pingcap/pd/pull/622

> Make send and recv in two different goroutines, so processTSORequests doesn't need
> to wait send-recv network latency and response to channel events quickly.
> 
> This strategy is trade throughput for latency, essentially.

Benchmark:

* original => 80w QPS, pd CPU 10%

```
count:929387, max:5, min:0, >1ms:383047, >2ms:20713, >5ms:58, >10ms:0, >30ms:0
count:918670, max:5, min:0, >1ms:374574, >2ms:30030, >5ms:27, >10ms:0, >30ms:0
count:901812, max:13, min:0, >1ms:399512, >2ms:33938, >5ms:9, >10ms:1, >30ms:0
count:912241, max:6, min:0, >1ms:398525, >2ms:28713, >5ms:1, >10ms:0, >30ms:0
count:903054, max:6, min:0, >1ms:386886, >2ms:28067, >5ms:392, >10ms:0, >30ms:0
count:912829, max:6, min:0, >1ms:344047, >2ms:45538, >5ms:1056, >10ms:0, >30ms:0
count:843504, max:8, min:0, >1ms:417298, >2ms:59063, >5ms:1156, >10ms:0, >30ms:0
count:811959, max:7, min:0, >1ms:512761, >2ms:52206, >5ms:565, >10ms:0, >30ms:0
count:796683, max:7, min:0, >1ms:555371, >2ms:39364, >5ms:310, >10ms:0, >30ms:0
count:757815, max:7, min:0, >1ms:525799, >2ms:62401, >5ms:1108, >10ms:0, >30ms:0
count:829600, max:7, min:0, >1ms:479063, >2ms:37831, >5ms:137, >10ms:0, >30ms:0
count:899534, max:7, min:0, >1ms:421662, >2ms:27072, >5ms:408, >10ms:0, >30ms:0
count:894057, max:5, min:0, >1ms:409253, >2ms:28615, >5ms:67, >10ms:0, >30ms:0
count:827200, max:7, min:0, >1ms:439732, >2ms:50011, >5ms:127, >10ms:0, >30ms:0
count:828589, max:8, min:0, >1ms:510693, >2ms:32011, >5ms:1002, >10ms:0, >30ms:0
count:781341, max:8, min:0, >1ms:557621, >2ms:40993, >5ms:673, >10ms:0, >30ms:0
count:785697, max:8, min:0, >1ms:560285, >2ms:37854, >5ms:355, >10ms:0, >30ms:0
count:741494, max:7, min:0, >1ms:550978, >2ms:58130, >5ms:1196, >10ms:0, >30ms:0
count:820620, max:7, min:0, >1ms:488421, >2ms:44165, >5ms:392, >10ms:0, >30ms:0
count:808598, max:8, min:0, >1ms:422706, >2ms:51282, >5ms:758, >10ms:0, >30ms:0
count:856351, max:9, min:0, >1ms:386263, >2ms:51499, >5ms:514, >10ms:0, >30ms:0
count:895857, max:6, min:0, >1ms:389978, >2ms:35260, >5ms:36, >10ms:0, >30ms:0
count:811253, max:8, min:0, >1ms:469976, >2ms:49290, >5ms:1071, >10ms:0, >30ms:0
count:789586, max:6, min:0, >1ms:533432, >2ms:46689, >5ms:159, >10ms:0, >30ms:0
count:783189, max:6, min:0, >1ms:530694, >2ms:50365, >5ms:914, >10ms:0, >30ms:0
count:801364, max:7, min:0, >1ms:539804, >2ms:45166, >5ms:779, >10ms:0, >30ms:0
count:821978, max:6, min:0, >1ms:485155, >2ms:42966, >5ms:966, >10ms:0, >30ms:0
count:896909, max:5, min:0, >1ms:435517, >2ms:29165, >5ms:17, >10ms:0, >30ms:0
count:869623, max:7, min:0, >1ms:410536, >2ms:35552, >5ms:324, >10ms:0, >30ms:0
```

* GetTSAsync without batch => 8w QPS,  pd CPU 150%

```
count:94694, max:22, min:0, >1ms:0, >2ms:1717, >5ms:38286, >10ms:54654, >30ms:0
count:83931, max:26, min:2, >1ms:0, >2ms:2277, >5ms:24997, >10ms:56657, >30ms:0
count:81082, max:26, min:2, >1ms:0, >2ms:1316, >5ms:19210, >10ms:60556, >30ms:0
count:81625, max:22, min:1, >1ms:52, >2ms:1066, >5ms:18526, >10ms:61981, >30ms:0
count:84306, max:24, min:3, >1ms:0, >2ms:1129, >5ms:22578, >10ms:60599, >30ms:0
count:88422, max:23, min:3, >1ms:0, >2ms:1391, >5ms:32398, >10ms:54633, >30ms:0
count:93769, max:22, min:2, >1ms:0, >2ms:3317, >5ms:38444, >10ms:52008, >30ms:0
count:89965, max:24, min:2, >1ms:0, >2ms:3641, >5ms:30087, >10ms:56237, >30ms:0
count:94207, max:18, min:2, >1ms:0, >2ms:1873, >5ms:38906, >10ms:53428, >30ms:0
count:87171, max:25, min:2, >1ms:0, >2ms:1184, >5ms:30741, >10ms:55246, >30ms:0
count:82392, max:27, min:2, >1ms:0, >2ms:1197, >5ms:22748, >10ms:58447, >30ms:0
count:83627, max:26, min:2, >1ms:0, >2ms:2013, >5ms:22077, >10ms:59537, >30ms:0
count:81564, max:24, min:2, >1ms:0, >2ms:1081, >5ms:19558, >10ms:60925, >30ms:0
count:76729, max:28, min:2, >1ms:0, >2ms:1352, >5ms:16251, >10ms:59126, >30ms:0
count:65841, max:32, min:4, >1ms:0, >2ms:35, >5ms:8335, >10ms:57351, >30ms:120
count:64438, max:26, min:4, >1ms:0, >2ms:10, >5ms:3613, >10ms:60815, >30ms:0
count:64358, max:26, min:4, >1ms:0, >2ms:30, >5ms:3082, >10ms:61246, >30ms:0
count:69790, max:27, min:4, >1ms:0, >2ms:148, >5ms:10146, >10ms:59496, >30ms:0
count:77491, max:24, min:2, >1ms:0, >2ms:485, >5ms:14475, >10ms:62531, >30ms:0
count:77892, max:26, min:2, >1ms:0, >2ms:714, >5ms:14842, >10ms:62336, >30ms:0
count:76488, max:25, min:3, >1ms:0, >2ms:598, >5ms:13454, >10ms:62436, >30ms:0
count:78819, max:23, min:2, >1ms:0, >2ms:1091, >5ms:15055, >10ms:62673, >30ms:0
count:82173, max:27, min:2, >1ms:0, >2ms:606, >5ms:20580, >10ms:60987, >30ms:0
count:82358, max:23, min:1, >1ms:25, >2ms:803, >5ms:17638, >10ms:63892, >30ms:0
count:83726, max:24, min:3, >1ms:0, >2ms:1624, >5ms:20545, >10ms:61557, >30ms:0
count:87301, max:21, min:2, >1ms:0, >2ms:1551, >5ms:24710, >10ms:61040, >30ms:0
```

* GetTSAsync + batch => 18w QPS,  pd CPU 110%

```
count:226967, max:18, min:0, >1ms:52589, >2ms:102910, >5ms:51520, >10ms:18196, >30ms:0
count:187118, max:21, min:0, >1ms:31955, >2ms:75641, >5ms:53819, >10ms:25010, >30ms:0
count:193912, max:24, min:0, >1ms:36677, >2ms:78632, >5ms:53564, >10ms:23218, >30ms:0
count:200349, max:22, min:0, >1ms:45058, >2ms:75687, >5ms:55802, >10ms:22308, >30ms:0
count:207106, max:25, min:0, >1ms:42533, >2ms:85472, >5ms:59705, >10ms:18015, >30ms:0
count:184643, max:20, min:0, >1ms:22104, >2ms:87632, >5ms:51873, >10ms:22438, >30ms:0
count:181453, max:24, min:0, >1ms:27566, >2ms:76822, >5ms:51147, >10ms:25411, >30ms:0
count:177724, max:28, min:0, >1ms:30822, >2ms:74606, >5ms:43257, >10ms:28225, >30ms:0
count:177316, max:27, min:0, >1ms:25849, >2ms:71631, >5ms:57031, >10ms:22509, >30ms:0
count:210723, max:18, min:0, >1ms:46440, >2ms:92235, >5ms:47099, >10ms:22867, >30ms:0
count:201395, max:29, min:0, >1ms:39910, >2ms:88794, >5ms:51282, >10ms:20211, >30ms:0
count:194324, max:23, min:0, >1ms:36471, >2ms:79072, >5ms:54644, >10ms:22528, >30ms:0
count:209035, max:26, min:0, >1ms:43472, >2ms:95726, >5ms:48025, >10ms:20107, >30ms:0
count:200948, max:28, min:0, >1ms:34513, >2ms:94210, >5ms:49180, >10ms:21837, >30ms:0
count:200517, max:19, min:0, >1ms:30911, >2ms:94909, >5ms:53152, >10ms:21128, >30ms:0
count:179066, max:22, min:0, >1ms:25723, >2ms:73457, >5ms:52051, >10ms:27178, >30ms:0
count:171036, max:24, min:0, >1ms:25224, >2ms:64739, >5ms:50137, >10ms:30057, >30ms:0
count:214986, max:27, min:0, >1ms:46098, >2ms:98936, >5ms:49189, >10ms:19633, >30ms:0
count:206222, max:21, min:0, >1ms:47965, >2ms:87075, >5ms:47587, >10ms:22119, >30ms:0
count:206949, max:25, min:0, >1ms:42918, >2ms:83855, >5ms:57034, >10ms:20974, >30ms:0
count:209988, max:21, min:0, >1ms:46039, >2ms:94372, >5ms:43864, >10ms:23447, >30ms:0
count:188582, max:20, min:0, >1ms:26794, >2ms:78138, >5ms:61382, >10ms:21110, >30ms:0
count:181479, max:30, min:0, >1ms:23146, >2ms:78029, >5ms:55305, >10ms:24386, >30ms:121
count:198205, max:27, min:0, >1ms:27889, >2ms:96630, >5ms:50786, >10ms:21837, >30ms:0
count:178834, max:25, min:0, >1ms:27213, >2ms:81467, >5ms:42947, >10ms:26310, >30ms:0
count:176256, max:40, min:0, >1ms:14391, >2ms:96344, >5ms:35371, >10ms:29563, >30ms:7
count:166028, max:26, min:0, >1ms:10613, >2ms:87432, >5ms:39900, >10ms:27772, >30ms:0
count:153483, max:28, min:0, >1ms:6900, >2ms:69609, >5ms:50339, >10ms:26627, >30ms:0
count:148129, max:29, min:0, >1ms:8887, >2ms:60586, >5ms:46563, >10ms:32058, >30ms:0
count:158489, max:27, min:0, >1ms:15377, >2ms:66453, >5ms:46446, >10ms:29853, >30ms:0
```

### CODE REVIEW

**Comment by siddontang** [comment]
seem origin still has the best performance, why do we not use it?

**Comment by siddontang** [comment]
can we close it now, @tiancaiamao ?

**Comment by tiancaiamao** [comment]
Wait a moment please, I want to test some variant of this code. @siddontang

**Comment by tiancaiamao** [comment]
After tune, this branch could reach 170w QPS, compare to 100w QPS of master branch,
and the latency is also lower than before.
I think it works!

PTAL @siddontang @disksing @coocood

**Comment by coocood** [comment]
@tiancaiamao 
More detailed performance result and CPU usage?

**Comment by tiancaiamao** [comment]
After:    pd-tso-bench CPU 665%, pd-server CPU 86%
```
➜  [115] bin git:(25224fe) ✗ ./pd-tso-bench -pd "192.168.199.118:2379"       
INFO[0000] [pd] create pd client with endpoints [192.168.199.118:2379] 
INFO[0000] [pd] leader switches to: http://192.168.199.118:2379, previous:  
INFO[0000] [pd] init cluster id 6426875904493309296     
count:1766663, max:4, min:0, >1ms:59320, >2ms:5817, >5ms:0, >10ms:0, >30ms:0
count:1731660, max:3, min:0, >1ms:71218, >2ms:4917, >5ms:0, >10ms:0, >30ms:0
count:1769959, max:2, min:0, >1ms:63565, >2ms:6289, >5ms:0, >10ms:0, >30ms:0
count:1747821, max:3, min:0, >1ms:63644, >2ms:7303, >5ms:0, >10ms:0, >30ms:0
count:1735035, max:2, min:0, >1ms:72543, >2ms:4677, >5ms:0, >10ms:0, >30ms:0
count:1763472, max:2, min:0, >1ms:71101, >2ms:1830, >5ms:0, >10ms:0, >30ms:0
count:1794247, max:2, min:0, >1ms:57649, >2ms:2506, >5ms:0, >10ms:0, >30ms:0
count:1755183, max:2, min:0, >1ms:81964, >2ms:1438, >5ms:0, >10ms:0, >30ms:0
count:1737065, max:2, min:0, >1ms:73377, >2ms:6844, >5ms:0, >10ms:0, >30ms:0
count:1788889, max:2, min:0, >1ms:65929, >2ms:1331, >5ms:0, >10ms:0, >30ms:0
count:1722003, max:3, min:0, >1ms:75876, >2ms:6213, >5ms:0, >10ms:0, >30ms:
```

Before:  pd-tso-bench CPU 428%, pd-server CPU 30%
```
➜  [115] bin git:(25224fe) ✗ ./pd-tso-bench -pd "192.168.199.118:2379"
INFO[0000] [pd] create pd client with endpoints [192.168.199.118:2379] 
INFO[0000] [pd] leader switches to: http://192.168.199.118:2379, previous:  
INFO[0000] [pd] init cluster id 6426875904493309296     
count:1015093, max:6, min:0, >1ms:442941, >2ms:3007, >5ms:283, >10ms:0, >30ms:0
count:983313, max:3, min:0, >1ms:479283, >2ms:1999, >5ms:0, >10ms:0, >30ms:0
count:977092, max:2, min:0, >1ms:510326, >2ms:3, >5ms:0, >10ms:0, >30ms:0
count:1076943, max:3, min:0, >1ms:328533, >2ms:1869, >5ms:0, >10ms:0, >30ms:0
count:1045777, max:5, min:0, >1ms:377137, >2ms:627, >5ms:1, >10ms:0, >30ms:0
count:1009418, max:3, min:0, >1ms:458588, >2ms:1505, >5ms:0, >10ms:0, >30ms:0
count:1008807, max:2, min:0, >1ms:469728, >2ms:62, >5ms:0, >10ms:0, >30ms:0
count:1007306, max:3, min:0, >1ms:470365, >2ms:1645, >5ms:0, >10ms:0, >30ms:0
count:1016277, max:3, min:0, >1ms:451836, >2ms:350, >5ms:0, >10ms:0, >30ms:0
count:1003122, max:3, min:0, >1ms:474957, >2ms:1053, >5ms:0, >10ms:0, >30ms:0
count:990887, max:2, min:0, >1ms:494182, >2ms:467, >5ms:0, >10ms:0, >30ms:0
count:970245, max:2, min:0, >1ms:526198, >2ms:351, >5ms:0, >10ms:0, >30ms:0
count:985610, max:4, min:0, >1ms:499713, >2ms:2467, >5ms:0, >10ms:0, >30ms:0
count:1002671, max:2, min:0, >1ms:470846, >2ms:306, >5ms:0, >10ms:0, >30ms:0
count:1043497, max:3, min:0, >1ms:399441, >2ms:1459, >5ms:0, >10ms:0, >30ms:0
count:1022783, max:1, min:0, >1ms:434168, >2ms:0, >5ms:0, >10ms:0, >30ms:0
count:988805, max:5, min:0, >1ms:489818, >2ms:1632, >5ms:1, >10ms:0, >30ms:0
count:1003523, max:2, min:0, >1ms:471856, >2ms:221, >5ms:0, >10ms:0, >30ms:0
count:1004084, max:2, min:0, >1ms:475264, >2ms:751, >5ms:0, >10ms:0, >30ms:0
count:984307, max:4, min:0, >1ms:498610, >2ms:1034, >5ms:0, >10ms:0, >30ms:0
```

@coocood

File: pd-client/client.go [hunk 1]
```diff
@@ -528,3 +537,175 @@ func addrsToUrls(addrs []string) []string {
 	}
 	return urls
 }
+
+func (c *client) pdTSOClient() pdpb.PD_TsoClient {
+	ctx := context.Background()
```
Comment (siddontang) [inline pd-client/client.go]: `WithCancel(c.ctx)` 

Don't use Background

**Comment by siddontang** [comment]
How to handle timeout?

**Comment by tiancaiamao** [comment]
Not yet.
This API is asynchronous so it won't make the caller block. @siddontang

**Comment by siddontang** [comment]
Other place also needs to call Wait, this can block.

**Comment by tiancaiamao** [comment]
Now pass a `ctx.WithTimeout()` would control the timeout.
PTAL @siddontang @disksing @coocood

**Comment by siddontang** [comment]
Any benchmark now? I worry too many timers will hurt the performance.

**Comment by tiancaiamao** [comment]
WithTimeout() slight affect the result, but I think it's still acceptable.
CPU goes from 665% to 695%.

```
➜  [115] bin git:(00efc37) ✗ ./pd-tso-bench -pd "192.168.199.118:2379"        
INFO[0000] [pd] create pd client with endpoints [192.168.199.118:2379] 
INFO[0000] [pd] leader switches to: http://192.168.199.118:2379, previous:  
INFO[0000] [pd] init cluster id 6426875904493309296     
count:1514209, max:9, min:0, >1ms:177816, >2ms:22633, >5ms:671, >10ms:0, >30ms:0
count:1556232, max:3, min:0, >1ms:157221, >2ms:20917, >5ms:0, >10ms:0, >30ms:0
count:1610625, max:2, min:0, >1ms:158271, >2ms:10017, >5ms:0, >10ms:0, >30ms:0
count:1585037, max:4, min:0, >1ms:155283, >2ms:21495, >5ms:0, >10ms:0, >30ms:0
count:1589266, max:3, min:0, >1ms:164893, >2ms:10394, >5ms:0, >10ms:0, >30ms:0
count:1537842, max:4, min:0, >1ms:171930, >2ms:22383, >5ms:0, >10ms:0, >30ms:0
count:1550060, max:3, min:0, >1ms:173469, >2ms:20129, >5ms:0, >10ms:0, >30ms:0
count:1554103, max:4, min:0, >1ms:159501, >2ms:18023, >5ms:0, >10ms:0, >30ms:0
count:1544470, max:4, min:0, >1ms:165368, >2ms:18989, >5ms:0, >10ms:0, >30ms:0
count:1562577, max:3, min:0, >1ms:172731, >2ms:16997, >5ms:0, >10ms:0, >30ms:0
count:1554627, max:3, min:0, >1ms:165342, >2ms:16671, >5ms:0, >10ms:0, >30ms:0
count:1529211, max:7, min:0, >1ms:164869, >2ms:20365, >5ms:999, >10ms:0, >30ms:0
count:1533072, max:3, min:0, >1ms:180492, >2ms:13687, >5ms:0, >10ms:0, >30ms:0
```

@siddontang

**Comment by siddontang** [comment]
But QPS reduces from 1722003 to 1533072, about 10%.

**Comment by tiancaiamao** [comment]
Through timeout context slightly hurt the performance, it still beat the request-response way.

1. In the benchmark battle, new branch shows better QPS, and with a larger percentage latency < 1ms.
2. In the real world case, there won't have that much pressure, so the request-response way batch count reduce and perform poor, while streaming way won't be affected.

If I test with 500 connection, and sleep 5ms, the result is as follows:

Before:
```
➜  [115] bin git:(00efc37) ✗ ./pd-tso-bench -pd "192.168.199.118:2379" -C 500 -sleep 5ms
INFO[0000] [pd] create pd client with endpoints [192.168.199.118:2379] 
INFO[0000] [pd] leader switches to: http://192.168.199.118:2379, previous:  
INFO[0000] [pd] init cluster id 6426875904493309296     
count:574993, max:2, min:0, >1ms:106863, >2ms:427, >5ms:0, >10ms:0, >30ms:0
count:609093, max:2, min:0, >1ms:60458, >2ms:157, >5ms:0, >10ms:0, >30ms:0
count:628347, max:3, min:0, >1ms:89517, >2ms:4, >5ms:0, >10ms:0, >30ms:0
count:746333, max:2, min:0, >1ms:1377, >2ms:171, >5ms:0, >10ms:0, >30ms:0
count:665922, max:2, min:0, >1ms:25801, >2ms:97, >5ms:0, >10ms:0, >30ms:0
count:635950, max:1, min:0, >1ms:22853, >2ms:0, >5ms:0, >10ms:0, >30ms:0
count:631592, max:3, min:0, >1ms:31806, >2ms:207, >5ms:0, >10ms:0, >30ms:0
count:607635, max:2, min:0, >1ms:59624, >2ms:492, >5ms:0, >10ms:0, >30ms:0
count:599039, max:1, min:0, >1ms:66919, >2ms:0, >5ms:0, >10ms:0, >30ms:0
count:627326, max:2, min:0, >1ms:31910, >2ms:259, >5ms:0, >10ms:0, >30ms:0
count:622266, max:2, min:0, >1ms:32692, >2ms:103, >5ms:0, >10ms:0, >30ms:0
count:591833, max:1, min:0, >1ms:70430, >2ms:0, >5ms:0, >10ms:0, >30ms:0
count:598753, max:3, min:0, >1ms:63633, >2ms:502, >5ms:0, >10ms:0, >30ms:0
count:592854, max:2, min:0, >1ms:76142, >2ms:39, >5ms:0, >10ms:0, >30ms:0
count:591635, max:3, min:0, >1ms:71246, >2ms:396, >5ms:0, >10ms:0, >30ms:0
count:590579, max:2, min:0, >1ms:63295, >2ms:619, >5ms:0, >10ms:0, >30ms:0
```

After:
```
➜  [115] bin git:(00efc37) ✗ ./pd-tso-bench -pd "192.168.199.118:2379" -C 500 -sleep 5ms 
INFO[0000] [pd] create pd client with endpoints [192.168.199.118:2379] 
INFO[0000] [pd] leader switches to: http://192.168.199.118:2379, previous:  
INFO[0000] [pd] init cluster id 6426875904493309296     
count:1251201, max:4, min:0, >1ms:74651, >2ms:1554, >5ms:0, >10ms:0, >30ms:0
count:1156134, max:2, min:0, >1ms:83213, >2ms:1902, >5ms:0, >10ms:0, >30ms:0
count:1258752, max:4, min:0, >1ms:71080, >2ms:549, >5ms:0, >10ms:0, >30ms:0
count:1201618, max:2, min:0, >1ms:85059, >2ms:1287, >5ms:0, >10ms:0, >30ms:0
count:1243404, max:5, min:0, >1ms:79653, >2ms:1940, >5ms:83, >10ms:0, >30ms:0
count:1261740, max:2, min:0, >1ms:73828, >2ms:469, >5ms:0, >10ms:0, >30ms:0
count:1267599, max:2, min:0, >1ms:80703, >2ms:376, >5ms:0, >10ms:0, >30ms:0
count:1228513, max:2, min:0, >1ms:89393, >2ms:1252, >5ms:0, >10ms:0, >30ms:0
count:1239034, max:2, min:0, >1ms:82583, >2ms:2877, >5ms:0, >10ms:0, >30ms:0
count:1193033, max:2, min:0, >1ms:93077, >2ms:2589, >5ms:0, >10ms:0, >30ms:0
count:1242565, max:3, min:0, >1ms:84474, >2ms:1799, >5ms:0, >10ms:0, >30ms:0
count:1208362, max:2, min:0, >1ms:84238, >2ms:1262, >5ms:0, >10ms:0, >30ms:0
count:1156218, max:2, min:0, >1ms:93749, >2ms:1702, >5ms:0, >10ms:0, >30ms:0
count:1169839, max:4, min:0, >1ms:90786, >2ms:5344, >5ms:0, >10ms:0, >30ms:0
count:1198327, max:2, min:0, >1ms:91646, >2ms:4367, >5ms:0, >10ms:0, >30ms:0
count:1224089, max:2, min:0, >1ms:90332, >2ms:1823, >5ms:0, >10ms:0, >30ms:0
count:1233253, max:2, min:0, >1ms:94360, >2ms:717, >5ms:0, >10ms:0, >30ms:0
```

**Comment by tiancaiamao** [comment]
Test in TiDB, no improvement.
So I close it.

**Comment by coocood** [comment]
@tiancaiamao 
How did you test it in TiDB?
Why close this PR if it has better performance?

If the network latency between TiDB and PD is high, waiting for previous response for new request has unnecessary latency cost.

**Pull request closed without merging**